File: FSKit/Classes/UIAdapter/FSUIAdapter.m
////
////  HEUIAdapter.m
////  Pods
////
////  Created by 扶冬冬 on 2021/2/1.
////
//
//#import "FSUIAdapter.h"
//
//@implementation FSUIAdapter {
//    CGFloat     _originRatioForIPhone;
//    CGFloat     _originRatioForIPad;
//}
//
//+ (FSUIAdapter *)sharedInstance {
//    static FSUIAdapter *_instance = nil;
//    static dispatch_once_t onceToken;
//    dispatch_once(&onceToken, ^{
//        _instance = [[FSUIAdapter alloc] init];
//    });
//    return _instance;
//}
//
//- (instancetype)init {
//    self = [super init];
//    if (self) {
//        _isIPad = [UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPad;
//        _isIPhone = [UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPhone;
//        if (UIScreen.mainScreen.bounds.size.width > UIScreen.mainScreen.bounds.size.height) {
//            _screenSmallerValue = UIScreen.mainScreen.bounds.size.height;
//            _screenbiggerValue = UIScreen.mainScreen.bounds.size.width;
//        } else {
//            _screenSmallerValue = UIScreen.mainScreen.bounds.size.width;
//            _screenbiggerValue = UIScreen.mainScreen.bounds.size.height;
//        }
//        _originRatioForIPhone = _screenSmallerValue / 375.0;
//        _originRatioForIPad   = _screenSmallerValue / 810.0;
//        
//        _model1p6 = [self modelForRatio:1.6];
//        _model1p3 = [self modelForRatio:1.3];
//        
//        if (_isIPad) {
//            _iPadValue = _model1p6;
//        }
//
//        [self calculateScreenHeightRatios];
//    }
//    return self;
//}
//
//- (FSUIAdapterModel *)model {
//    if (!_model) {
//        _model = [self modelForRatio:2.16];
//    }
//    return _model;
//}
//
//- (FSUIAdapterModel *)model1p88 {
//    if (!_model1p88) {
//        _model1p88 = [self modelForRatio:1.88];
//    }
//    return _model1p88;
//}
//
//- (FSUIAdapterModel *)model1p75 {
//    if (!_model1p75) {
//        _model1p75 = [self modelForRatio:1.75];
//    }
//    return _model1p75;
//}
//
//- (FSUIAdapterModel *)model1p1 {
//    if (!_model1p1) {
//        _model1p1 = [self modelForRatio:1.135];
//    }
//    return _model1p1;
//}
//
//- (FSUIAdapterModel *)modelForRatio:(CGFloat)iPadToIPhone {
//    CGFloat ratio = _originRatioForIPhone;
//    if (_isIPad) {
//        ratio = _originRatioForIPad * iPadToIPhone;
//    }
//    FSUIAdapterModel *model = [[FSUIAdapterModel alloc] initWithRatio:ratio];
//    return model;
//}
//
//- (void)calculateScreenHeightRatios {
//    _height05 = ceil(_screenSmallerValue * .05);
//    _height10 = ceil(_screenSmallerValue * .10);
//    _height15 = ceil(_screenSmallerValue * .15);
//    _height20 = ceil(_screenSmallerValue * .20);
//    _height25 = ceil(_screenSmallerValue * .25);
//    _height30 = ceil(_screenSmallerValue * .30);
//    _height35 = ceil(_screenSmallerValue * .35);
//    _height40 = ceil(_screenSmallerValue * .40);
//    _height45 = ceil(_screenSmallerValue * .45);
//    _height50 = ceil(_screenSmallerValue * .50);
//    _height55 = ceil(_screenSmallerValue * .55);
//    _height60 = ceil(_screenSmallerValue * .60);
//    _height65 = ceil(_screenSmallerValue * .65);
//    _height70 = ceil(_screenSmallerValue * .70);
//    _height75 = ceil(_screenSmallerValue * .75);
//    _height80 = ceil(_screenSmallerValue * .80);
//    _height85 = ceil(_screenSmallerValue * .85);
//    _height90 = ceil(_screenSmallerValue * .90);
//    _height95 = ceil(_screenSmallerValue * .95);
//    
//    _width05 = ceil(_screenbiggerValue * .05);
//    _width10 = ceil(_screenbiggerValue * .10);
//    _width15 = ceil(_screenbiggerValue * .15);
//    _width20 = ceil(_screenbiggerValue * .20);
//    _width25 = ceil(_screenbiggerValue * .25);
//    _width30 = ceil(_screenbiggerValue * .30);
//    _width35 = ceil(_screenbiggerValue * .35);
//    _width40 = ceil(_screenbiggerValue * .40);
//    _width45 = ceil(_screenbiggerValue * .45);
//    _width50 = ceil(_screenbiggerValue * .50);
//    _width55 = ceil(_screenbiggerValue * .55);
//    _width60 = ceil(_screenbiggerValue * .60);
//    _width65 = ceil(_screenbiggerValue * .65);
//    _width70 = ceil(_screenbiggerValue * .70);
//    _width75 = ceil(_screenbiggerValue * .75);
//    _width80 = ceil(_screenbiggerValue * .80);
//    _width85 = ceil(_screenbiggerValue * .85);
//    _width90 = ceil(_screenbiggerValue * .90);
//    _width95 = ceil(_screenbiggerValue * .95);
//}
//
//CGFloat _differentValueForIPad(CGFloat iPad, CGFloat iPhone) {
//    if (FSUIAdapter.sharedInstance.isIPad == true) {
//        return iPad;
//    }
//    return iPhone;
//}
//
//- (void)setSafeInsets:(UIEdgeInsets)safeInsets {
//    _safeInsets = safeInsets;
//    
//    _safeWidth = UIScreenWidth - safeInsets.left - safeInsets.right;
//    _safeHeight = UIScreenHeight - safeInsets.top - safeInsets.bottom;
//}
//
//@end
//
//@implementation UIView (Util)
//
//- (CGFloat)height {
//    return self.frame.size.height;
//}
//
//- (void)setHeight: (CGFloat) newheight {
//    CGRect newframe = self.frame;
//    newframe.size.height = newheight;
//    self.frame = newframe;
//}
//
//- (CGFloat)width {
//    return self.frame.size.width;
//}
//
//- (void)setWidth:(CGFloat)newwidth {
//    CGRect newframe = self.frame;
//    newframe.size.width = newwidth;
//    self.frame = newframe;
//}
//
//- (CGFloat)top {
//    return self.frame.origin.y;
//}
//
//- (void)setTop:(CGFloat)newtop {
//    CGRect newframe = self.frame;
//    newframe.origin.y = newtop;
//    self.frame = newframe;
//}
//
//- (CGFloat)left {
//    return self.frame.origin.x;
//}
//
//- (void)setLeft:(CGFloat)newleft {
//    CGRect newframe = self.frame;
//    newframe.origin.x = newleft;
//    self.frame = newframe;
//}
//
//- (CGFloat)bottom {
//    return self.frame.origin.y + self.frame.size.height;
//}
//
//- (void)setBottom:(CGFloat)newbottom {
//    CGRect newframe = self.frame;
//    newframe.origin.y = newbottom - self.frame.size.height;
//    self.frame = newframe;
//}
//
//- (CGFloat)right {
//    return self.frame.origin.x + self.frame.size.width;
//}
//
//- (void)setRight:(CGFloat)newright {
//    CGFloat delta = newright - (self.frame.origin.x + self.frame.size.width);
//    CGRect newframe = self.frame;
//    newframe.origin.x += delta ;
//    self.frame = newframe;
//}
//
//- (void)radiusWithRadius:(CGFloat)radius corner:(UIRectCorner)corner {
//    if (@available(iOS 11.0, *)) {
//        self.layer.cornerRadius = radius;
//        self.layer.maskedCorners = (CACornerMask)corner;
//    } else {
//        UIBezierPath * path = [UIBezierPath bezierPathWithRoundedRect:self.bounds byRoundingCorners:corner cornerRadii:CGSizeMake(radius, radius)];
//        CAShapeLayer *maskLayer = [[CAShapeLayer alloc] init];
//        maskLayer.frame = self.bounds;
//        maskLayer.path = path.CGPath;
//        self.layer.mask = maskLayer;
//    }
//}
//
//- (void)shadowColor:(UIColor *)shadowColor shadowOffset:(CGSize)size opacity:(CGFloat)opacity {
//    UIBezierPath *shadowPath = [UIBezierPath bezierPathWithRect:self.bounds];
//    self.layer.masksToBounds = NO;
//    if ([shadowColor isKindOfClass:UIColor.class]) {
//        self.layer.shadowColor = shadowColor.CGColor;
//    }
//    self.layer.shadowOffset = size;
//    self.layer.shadowOpacity = opacity;
//    self.layer.shadowRadius = self.layer.cornerRadius;
//    self.layer.shadowPath = shadowPath.CGPath;
//}
//
//- (void)cornerWithLeftTop:(CGFloat)leftTop
//                 rightTop:(CGFloat)rigtTop
//               bottomLeft:(CGFloat)bottemLeft
//              bottomRight:(CGFloat)bottemRight
//                    frame:(CGRect)frame {
//    
//    CGFloat width = frame.size.width;
//    CGFloat height = frame.size.height;
//    
//    UIBezierPath *maskPath = [UIBezierPath bezierPath];
//    maskPath.lineWidth = 1.0;
//    maskPath.lineCapStyle = kCGLineCapRound;
//    maskPath.lineJoinStyle = kCGLineJoinRound;
//    [maskPath moveToPoint:CGPointMake(bottemRight, height)]; //左下角
//    [maskPath addLineToPoint:CGPointMake(width - bottemRight, height)];
//    
//    [maskPath addQuadCurveToPoint:CGPointMake(width, height- bottemRight) controlPoint:CGPointMake(width, height)]; //右下角的圆弧
//    [maskPath addLineToPoint:CGPointMake(width, rigtTop)]; //右边直线
//    
//    [maskPath addQuadCurveToPoint:CGPointMake(width - rigtTop, 0) controlPoint:CGPointMake(width, 0)]; //右上角圆弧
//    [maskPath addLineToPoint:CGPointMake(leftTop, 0)]; //顶部直线
//    
//    [maskPath addQuadCurveToPoint:CGPointMake(0, leftTop) controlPoint:CGPointMake(0, 0)]; //左上角圆弧
//    [maskPath addLineToPoint:CGPointMake(0, height - bottemLeft)]; //左边直线
//    [maskPath addQuadCurveToPoint:CGPointMake(bottemLeft, height) controlPoint:CGPointMake(0, height)]; //左下角圆弧
//
//    CAShapeLayer *maskLayer = [CAShapeLayer layer];
//    maskLayer.frame = frame;
//    maskLayer.path = maskPath.CGPath;
//    maskLayer.frame = frame;
//    maskLayer.path = maskPath.CGPath;
//    self.layer.mask = maskLayer;
//}
//
//+ (CAGradientLayer *)gradientView:(UIView *)view frame:(CGRect)frame startPoint:(CGPoint)startPoint endPoint:(CGPoint)endPoint colors:(NSArray *)colors locations:(NSArray *)locations {
//    CAGradientLayer *gl = [CAGradientLayer layer];
//    gl.frame = frame;
//    gl.startPoint = startPoint;
//    gl.endPoint = endPoint;
//    gl.colors = colors;
//    gl.locations = locations;
////    [view.layer addSublayer:gl];
//    [view.layer insertSublayer:gl atIndex:0];
//    return gl;
//}
//
////- (UIEdgeInsets)safeAreaInsets {
////    if (@available(iOS 11.0, *)) {
////        return self.safeAreaInsets;
////    } else {
////        return UIEdgeInsetsZero;
////    }
////}
//
//@end

File: FSKit/Classes/FSUIAdapter.swift
//
//  FSUIAdapterCore.swift
//  FSKit
//
//  Translated from Objective-C to Swift
//

import UIKit

// MARK: - Convenience Accessors
public var UIScreenLong: CGFloat { FSUIAdapter.shared.screenBiggerValue }
public var UIScreenShort: CGFloat { FSUIAdapter.shared.screenSmallerValue }
public var UIScreenHeight: CGFloat { UIScreen.main.bounds.height }
public var UIScreenWidth: CGFloat { UIScreen.main.bounds.width }

// MARK: - FSUIAdapterModel
public class FSUIAdapterModel: NSObject {
    
    public let value0p5: CGFloat
    public let value1: CGFloat
    public let value2: CGFloat
    public let value3: CGFloat
    public let value4: CGFloat
    public let value5: CGFloat
    public let value6: CGFloat
    public let value7: CGFloat
    public let value7p5: CGFloat
    public let value8: CGFloat
    public let value9: CGFloat
    public let value10: CGFloat
    public let value11: CGFloat
    public let value12: CGFloat
    public let value13: CGFloat
    public let value14: CGFloat
    public let value15: CGFloat
    public let value16: CGFloat
    public let value17: CGFloat
    public let value18: CGFloat
    public let value19: CGFloat
    public let value20: CGFloat
    public let value22: CGFloat
    public let value25: CGFloat
    public let value26: CGFloat
    public let value30: CGFloat
    public let value36: CGFloat
    public let value38: CGFloat
    public let value40: CGFloat
    public let value45: CGFloat
    public let value46: CGFloat
    public let value47: CGFloat
    public let value50: CGFloat
    public let value55: CGFloat
    public let value60: CGFloat
    public let value65: CGFloat
    public let value70: CGFloat
    public let value75: CGFloat
    public let value80: CGFloat
    public let value85: CGFloat
    public let value90: CGFloat
    public let value95: CGFloat
    public let value100: CGFloat
    public let value105: CGFloat
    public let value110: CGFloat
    public let value120: CGFloat
    public let value125: CGFloat
    public let value130: CGFloat
    public let value135: CGFloat
    public let value140: CGFloat
    public let value145: CGFloat
    public let value150: CGFloat
    public let value160: CGFloat
    public let value165: CGFloat
    public let value170: CGFloat
    public let value180: CGFloat
    public let value190: CGFloat
    public let value200: CGFloat
    public let value210: CGFloat
    public let value230: CGFloat
    public let value250: CGFloat
    public let value260: CGFloat
    public let value270: CGFloat
    public let value280: CGFloat
    public let value290: CGFloat
    public let value300: CGFloat
    public let value310: CGFloat
    public let value320: CGFloat
    public let value330: CGFloat
    public let value340: CGFloat
    public let value350: CGFloat
    public let value375: CGFloat
    public let value400: CGFloat
    public let value500: CGFloat
    public let value650: CGFloat
    
    public init(ratio: CGFloat) {
        value0p5 = 0.5 * ratio
        value1 = 1 * ratio
        value2 = 2 * ratio
        value3 = 3 * ratio
        value4 = 4 * ratio
        value5 = 5 * ratio
        value6 = 6 * ratio
        value7 = 7 * ratio
        value7p5 = 7.5 * ratio
        value8 = 8 * ratio
        value9 = 9 * ratio
        value10 = ceil(10 * ratio)
        value11 = ceil(11 * ratio)
        value12 = ceil(12 * ratio)
        value13 = ceil(13 * ratio)
        value14 = ceil(14 * ratio)
        value15 = ceil(15 * ratio)
        value16 = ceil(16 * ratio)
        value17 = ceil(17 * ratio)
        value18 = ceil(18 * ratio)
        value19 = ceil(19 * ratio)
        value20 = ceil(20 * ratio)
        value22 = ceil(22 * ratio)
        value25 = ceil(25 * ratio)
        value26 = ceil(26 * ratio)
        value30 = ceil(30 * ratio)
        value36 = ceil(36 * ratio)
        value38 = ceil(38 * ratio)
        value40 = ceil(40 * ratio)
        value45 = ceil(45 * ratio)
        value46 = ceil(46 * ratio)
        value47 = ceil(47 * ratio)
        value50 = ceil(50 * ratio)
        value55 = ceil(55 * ratio)
        value60 = ceil(60 * ratio)
        value65 = ceil(65 * ratio)
        value70 = ceil(70 * ratio)
        value75 = ceil(75 * ratio)
        value80 = ceil(80 * ratio)
        value85 = ceil(85 * ratio)
        value90 = ceil(90 * ratio)
        value95 = ceil(95 * ratio)
        value100 = ceil(100 * ratio)
        value105 = ceil(105 * ratio)
        value110 = ceil(110 * ratio)
        value120 = ceil(120 * ratio)
        value125 = ceil(125 * ratio)
        value130 = ceil(130 * ratio)
        value135 = ceil(135 * ratio)
        value140 = ceil(140 * ratio)
        value145 = ceil(145 * ratio)
        value150 = ceil(150 * ratio)
        value160 = ceil(160 * ratio)
        value165 = ceil(165 * ratio)
        value170 = ceil(170 * ratio)
        value180 = ceil(180 * ratio)
        value190 = ceil(190 * ratio)
        value200 = ceil(200 * ratio)
        value210 = ceil(210 * ratio)
        value230 = ceil(230 * ratio)
        value250 = ceil(250 * ratio)
        value260 = ceil(260 * ratio)
        value270 = ceil(270 * ratio)
        value280 = ceil(280 * ratio)
        value290 = ceil(290 * ratio)
        value300 = ceil(300 * ratio)
        value310 = ceil(310 * ratio)
        value320 = ceil(320 * ratio)
        value330 = ceil(330 * ratio)
        value340 = ceil(340 * ratio)
        value350 = ceil(350 * ratio)
        value375 = ceil(375 * ratio)
        value400 = ceil(400 * ratio)
        value500 = ceil(500 * ratio)
        value650 = ceil(650 * ratio)
        super.init()
    }
}


// MARK: - FSUIAdapterManager
public class FSUIAdapter: NSObject {
    
    public static let shared = FSUIAdapter()
    
    public let isIPad: Bool
    public let isIPhone: Bool
    public let screenSmallerValue: CGFloat
    public let screenBiggerValue: CGFloat
    
    private let originRatioForIPhone: CGFloat
    private let originRatioForIPad: CGFloat
    
    public private(set) lazy var model: FSUIAdapterModel = modelForRatio(2.16)
    public private(set) lazy var model1p88: FSUIAdapterModel = modelForRatio(1.88)
    public private(set) lazy var model1p75: FSUIAdapterModel = modelForRatio(1.75)
    public private(set) lazy var model1p6: FSUIAdapterModel = modelForRatio(1.6)
    public private(set) lazy var model1p3: FSUIAdapterModel = modelForRatio(1.3)
    public private(set) lazy var model1p1: FSUIAdapterModel = modelForRatio(1.135)
    
    public var iPadValue: FSUIAdapterModel? {
        return isIPad ? model1p6 : nil
    }
    
    // Screen Height Ratios
    public private(set) var height05: CGFloat = 0
    public private(set) var height10: CGFloat = 0
    public private(set) var height15: CGFloat = 0
    public private(set) var height20: CGFloat = 0
    public private(set) var height25: CGFloat = 0
    public private(set) var height30: CGFloat = 0
    public private(set) var height35: CGFloat = 0
    public private(set) var height40: CGFloat = 0
    public private(set) var height45: CGFloat = 0
    public private(set) var height50: CGFloat = 0
    public private(set) var height55: CGFloat = 0
    public private(set) var height60: CGFloat = 0
    public private(set) var height65: CGFloat = 0
    public private(set) var height70: CGFloat = 0
    public private(set) var height75: CGFloat = 0
    public private(set) var height80: CGFloat = 0
    public private(set) var height85: CGFloat = 0
    public private(set) var height90: CGFloat = 0
    public private(set) var height95: CGFloat = 0
    
    // Screen Width Ratios
    public private(set) var width05: CGFloat = 0
    public private(set) var width10: CGFloat = 0
    public private(set) var width15: CGFloat = 0
    public private(set) var width20: CGFloat = 0
    public private(set) var width25: CGFloat = 0
    public private(set) var width30: CGFloat = 0
    public private(set) var width35: CGFloat = 0
    public private(set) var width40: CGFloat = 0
    public private(set) var width45: CGFloat = 0
    public private(set) var width50: CGFloat = 0
    public private(set) var width55: CGFloat = 0
    public private(set) var width60: CGFloat = 0
    public private(set) var width65: CGFloat = 0
    public private(set) var width70: CGFloat = 0
    public private(set) var width75: CGFloat = 0
    public private(set) var width80: CGFloat = 0
    public private(set) var width85: CGFloat = 0
    public private(set) var width90: CGFloat = 0
    public private(set) var width95: CGFloat = 0
    
    // Safe Area
    public var safeInsets: UIEdgeInsets = .zero {
        didSet {
            safeWidth = UIScreenWidth - safeInsets.left - safeInsets.right
            safeHeight = UIScreenHeight - safeInsets.top - safeInsets.bottom
        }
    }
    public private(set) var safeWidth: CGFloat = 0
    public private(set) var safeHeight: CGFloat = 0
    
    private override init() {
        isIPad = UIDevice.current.userInterfaceIdiom == .pad
        isIPhone = UIDevice.current.userInterfaceIdiom == .phone
        
        let screenBounds = UIScreen.main.bounds
        if screenBounds.width > screenBounds.height {
            screenSmallerValue = screenBounds.height
            screenBiggerValue = screenBounds.width
        } else {
            screenSmallerValue = screenBounds.width
            screenBiggerValue = screenBounds.height
        }
        
        originRatioForIPhone = screenSmallerValue / 375.0
        originRatioForIPad = screenSmallerValue / 810.0
        
        super.init()
        calculateScreenHeightRatios()
    }
    
    private func modelForRatio(_ iPadToIPhone: CGFloat) -> FSUIAdapterModel {
        var ratio = originRatioForIPhone
        if isIPad {
            ratio = originRatioForIPad * iPadToIPhone
        }
        return FSUIAdapterModel(ratio: ratio)
    }
    
    private func calculateScreenHeightRatios() {
        height05 = ceil(screenSmallerValue * 0.05)
        height10 = ceil(screenSmallerValue * 0.10)
        height15 = ceil(screenSmallerValue * 0.15)
        height20 = ceil(screenSmallerValue * 0.20)
        height25 = ceil(screenSmallerValue * 0.25)
        height30 = ceil(screenSmallerValue * 0.30)
        height35 = ceil(screenSmallerValue * 0.35)
        height40 = ceil(screenSmallerValue * 0.40)
        height45 = ceil(screenSmallerValue * 0.45)
        height50 = ceil(screenSmallerValue * 0.50)
        height55 = ceil(screenSmallerValue * 0.55)
        height60 = ceil(screenSmallerValue * 0.60)
        height65 = ceil(screenSmallerValue * 0.65)
        height70 = ceil(screenSmallerValue * 0.70)
        height75 = ceil(screenSmallerValue * 0.75)
        height80 = ceil(screenSmallerValue * 0.80)
        height85 = ceil(screenSmallerValue * 0.85)
        height90 = ceil(screenSmallerValue * 0.90)
        height95 = ceil(screenSmallerValue * 0.95)
        
        width05 = ceil(screenBiggerValue * 0.05)
        width10 = ceil(screenBiggerValue * 0.10)
        width15 = ceil(screenBiggerValue * 0.15)
        width20 = ceil(screenBiggerValue * 0.20)
        width25 = ceil(screenBiggerValue * 0.25)
        width30 = ceil(screenBiggerValue * 0.30)
        width35 = ceil(screenBiggerValue * 0.35)
        width40 = ceil(screenBiggerValue * 0.40)
        width45 = ceil(screenBiggerValue * 0.45)
        width50 = ceil(screenBiggerValue * 0.50)
        width55 = ceil(screenBiggerValue * 0.55)
        width60 = ceil(screenBiggerValue * 0.60)
        width65 = ceil(screenBiggerValue * 0.65)
        width70 = ceil(screenBiggerValue * 0.70)
        width75 = ceil(screenBiggerValue * 0.75)
        width80 = ceil(screenBiggerValue * 0.80)
        width85 = ceil(screenBiggerValue * 0.85)
        width90 = ceil(screenBiggerValue * 0.90)
        width95 = ceil(screenBiggerValue * 0.95)
    }
}

// MARK: - Helper Function
public func differentValueForIPad(_ iPad: CGFloat, _ iPhone: CGFloat) -> CGFloat {
    return FSUIAdapter.shared.isIPad ? iPad : iPhone
}

// MARK: - UIView Extension
public extension UIView {
    
    var fs_height: CGFloat {
        get { frame.size.height }
        set {
            var newFrame = frame
            newFrame.size.height = newValue
            frame = newFrame
        }
    }
    
    var fs_width: CGFloat {
        get { frame.size.width }
        set {
            var newFrame = frame
            newFrame.size.width = newValue
            frame = newFrame
        }
    }
    
    var fs_top: CGFloat {
        get { frame.origin.y }
        set {
            var newFrame = frame
            newFrame.origin.y = newValue
            frame = newFrame
        }
    }
    
    var fs_left: CGFloat {
        get { frame.origin.x }
        set {
            var newFrame = frame
            newFrame.origin.x = newValue
            frame = newFrame
        }
    }
    
    var fs_bottom: CGFloat {
        get { frame.origin.y + frame.size.height }
        set {
            var newFrame = frame
            newFrame.origin.y = newValue - frame.size.height
            frame = newFrame
        }
    }
    
    var fs_right: CGFloat {
        get { frame.origin.x + frame.size.width }
        set {
            let delta = newValue - (frame.origin.x + frame.size.width)
            var newFrame = frame
            newFrame.origin.x += delta
            frame = newFrame
        }
    }
    
    func fs_radius(radius: CGFloat, corner: UIRectCorner) {
        if #available(iOS 11.0, *) {
            layer.cornerRadius = radius
            layer.maskedCorners = CACornerMask(rawValue: corner.rawValue)
        } else {
            let path = UIBezierPath(roundedRect: bounds, byRoundingCorners: corner, cornerRadii: CGSize(width: radius, height: radius))
            let maskLayer = CAShapeLayer()
            maskLayer.frame = bounds
            maskLayer.path = path.cgPath
            layer.mask = maskLayer
        }
    }
    
    func fs_shadow(color: UIColor?, offset: CGSize, opacity: Float) {
        let shadowPath = UIBezierPath(rect: bounds)
        layer.masksToBounds = false
        if let color = color {
            layer.shadowColor = color.cgColor
        }
        layer.shadowOffset = offset
        layer.shadowOpacity = opacity
        layer.shadowRadius = layer.cornerRadius
        layer.shadowPath = shadowPath.cgPath
    }
    
    func fs_corner(leftTop: CGFloat, rightTop: CGFloat, bottomLeft: CGFloat, bottomRight: CGFloat, frame: CGRect) {
        let width = frame.size.width
        let height = frame.size.height
        
        let maskPath = UIBezierPath()
        maskPath.lineWidth = 1.0
        maskPath.lineCapStyle = .round
        maskPath.lineJoinStyle = .round
        
        maskPath.move(to: CGPoint(x: bottomRight, y: height))
        maskPath.addLine(to: CGPoint(x: width - bottomRight, y: height))
        maskPath.addQuadCurve(to: CGPoint(x: width, y: height - bottomRight), controlPoint: CGPoint(x: width, y: height))
        maskPath.addLine(to: CGPoint(x: width, y: rightTop))
        maskPath.addQuadCurve(to: CGPoint(x: width - rightTop, y: 0), controlPoint: CGPoint(x: width, y: 0))
        maskPath.addLine(to: CGPoint(x: leftTop, y: 0))
        maskPath.addQuadCurve(to: CGPoint(x: 0, y: leftTop), controlPoint: .zero)
        maskPath.addLine(to: CGPoint(x: 0, y: height - bottomLeft))
        maskPath.addQuadCurve(to: CGPoint(x: bottomLeft, y: height), controlPoint: CGPoint(x: 0, y: height))
        
        let maskLayer = CAShapeLayer()
        maskLayer.frame = frame
        maskLayer.path = maskPath.cgPath
        layer.mask = maskLayer
    }
    
    @discardableResult
    static func fs_gradient(view: UIView, frame: CGRect, startPoint: CGPoint, endPoint: CGPoint, colors: [Any], locations: [NSNumber]) -> CAGradientLayer {
        let gradientLayer = CAGradientLayer()
        gradientLayer.frame = frame
        gradientLayer.startPoint = startPoint
        gradientLayer.endPoint = endPoint
        gradientLayer.colors = colors
        gradientLayer.locations = locations
        view.layer.insertSublayer(gradientLayer, at: 0)
        return gradientLayer
    }
}
